### PR TITLE
Access Control Documentation

### PR DESCRIPTION
**Pull request by SilentCicero**
- Added access control documentation
- Added example

Closes: https://github.com/FuelLabs/sway/issues/1297

### CODE REVIEW

File: docs/src/blockchain-development/access_control.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# Access Control
+
+Smart contracts require the ability to restrict access to and identify certain users or contracts.
+
+Unlike account based blockchains, Fuel calls can originate from an origin which has potentially multiple parties.
+
+## msg_sender
+
+To deliver an akin experience to Ethereum based access control, the `std` library provides a `msg_sender` method, which determines a single sender based upon the call and transaction input data.
```
Comment (nfurfaro) [inline docs/src/blockchain-development/access_control.md:9]: "To deliver an  experience akin to Ethereum ..."

File: examples/msg_sender/src/main.sw [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    chain::auth::{AuthError, Sender, msg_sender}, 
+    panic::panic, 
+    result::*
+};
+
+abi MyOwnedContract {
+    fn receive(field_1: u64) -> bool;
+}
+
+const OWNER:b256 = 0x9ae5b658754e096e4d681c548daf46354495a437cc61492599e33fc64dcdc30c;
```
Comment (nfurfaro) [inline examples/msg_sender/src/main.sw:15]: soonish (tm) we'll be able to do
`const OWNER = ~Address::from(0x9ae5b658754e096e4d681c548daf46354495a437cc61492599e33fc64dcdc30c);`
and then below:
`assert(addr == OWNER);`
but the way you have it works with the current state of things.

**Review by nfurfaro — COMMENTED** [review]
Just 1 little thing, then looks good.

File: docs/src/blockchain-development/access_control.md (hunk 1 above)
Comment (SilentCicero) [reply docs/src/blockchain-development/access_control.md:9]: That sounds better :)

**Review by nfurfaro — APPROVED** [review]
LG.

File: docs/src/blockchain-development/access_control.md [hunk 3]
```diff
@@ -0,0 +1,18 @@
+# Access Control
+
+Smart contracts require the ability to restrict access to and identify certain users or contracts.
+
+Unlike account based blockchains, Fuel calls can originate from an origin which has potentially multiple parties.
```
Comment (adlerjohn) [inline docs/src/blockchain-development/access_control.md:5]:
```suggestion
Smart contracts require the ability to restrict access to and identify certain users or contracts. Unlike account-based blockchains, transactions in UTXO-based blockchains (i.e. Fuel) do not necessarily have a unique transaction sender. Additional logic is needed to handle this difference, and is provided by the standard library.
```

File: docs/src/blockchain-development/access_control.md [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+## msg_sender
```
Comment (adlerjohn) [inline docs/src/blockchain-development/access_control.md:7]:
```suggestion
## `msg_sender`
```

File: docs/src/blockchain-development/access_control.md [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+To deliver an experience akin to Ethereum based access control, the `std` library provides a `msg_sender` method, which determines a single sender based upon the call and transaction input data.
```
Comment (adlerjohn) [inline docs/src/blockchain-development/access_control.md:9]:
```suggestion
To deliver an experience akin to Ethereum's access control, the `std` library provides a `msg_sender` function, which identifies a unique caller based upon the call and/or transaction input data.
```

File: docs/src/blockchain-development/access_control.md [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+```sway
+{{#include ../../../examples/msg_sender/src/main.sw}}
+```
+
+The `msg_sender` method works as follows: 
+- If the caller is a contract, then `Result::Ok(Sender)` is returned with the `ContractId` specified.
+- If the caller is external (i.e. from a script), then all transaction input coin owners are checked, if all owners are the same, then `Result::Ok(Sender)` is returned with `Address` specified.
+- If input coin owners are different, then the `Sender` cannot be determined and an `Result::Err(AuthError)` is returned.
```
Comment (adlerjohn) [inline docs/src/blockchain-development/access_control.md:18]:
```suggestion
The `msg_sender` function works as follows:

- If the caller is a contract, then `Result::Ok(Sender)` is returned with the `ContractId` sender variant.
- If the caller is external (i.e. from a script), then all coin input owners in the transaction are checked. If all owners are the same, then `Result::Ok(Sender)` is returned with the `Address` sender variant.
- If the caller is external and coin input owners are different, then the caller cannot be determined and a `Result::Err(AuthError)` is returned.
```

File: examples/msg_sender/src/main.sw [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+        let sender: Result<Sender, AuthError> = msg_sender();
+        if let Sender::Address(addr) = sender.unwrap() {
+            assert(addr.into() == OWNER);
+        } else {
+            panic(0);
+        };
+
+        true
+    }
+}
```
Comment (adlerjohn) [inline examples/msg_sender/src/main.sw:28]: format this file

**Review by adlerjohn — CHANGES_REQUESTED** [review]

File: examples/msg_sender/src/main.sw (hunk 7 above)
Comment (SilentCicero) [reply examples/msg_sender/src/main.sw:28]: Will do.

**Review by adlerjohn — APPROVED** [review]

**Pull request merged**